File: MRiLab-GUILess/PreScan.m
function PreScan()

global VObj;
global VCtl;
global VMag;
global VCoi;
global VVar;
global VSig;

global VMmg;
global VMco;
global VMgd;

VCtl=[]; % update VCtl

% FOV & resolution
  % Cartesian
    VCtl.ResFreq = 100;
    VCtl.RFreq=0.002;
    VCtl.ResPhase = 80; % guarantee even number of Ky sample points for Ky = 0
    VCtl.RPhase=0.0025;
    VCtl.SliceNum = 1; % guarantee even number of Kz sample points for Kz = 0
    VCtl.RSlice=0.006;
    VCtl.FOVSlice=0.006;
    VCtl.TrajType='cartesian';
    VCtl.FirstPhNum = 80;
    VCtl.SecondPhNum = 1;
    %These are the calculations
%   VCtl.ResFreq = VCtl.ResFreq + ~mod(VCtl.ResFreq,2) % guarantee odd number of Kx sample points for sampling echo peak when Kx = 0
%   VCtl.RFreq=VCtl.FOVFreq/(VCtl.ResFreq - 1);
%   VCtl.ResPhase = VCtl.ResPhase - mod(VCtl.ResPhase,2); % guarantee even number of Ky sample points for Ky = 0
%   VCtl.RPhase=VCtl.FOVPhase/VCtl.ResPhase;
%   VCtl.SliceNum = max(1,VCtl.SliceNum - mod(VCtl.SliceNum,2)); % guarantee even number of Kz sample points for Kz = 0
%   VCtl.RSlice=VCtl.SliceThick;
%   VCtl.FOVSlice=VCtl.SliceNum*VCtl.SliceThick;
%   VCtl.TrajType='cartesian';
%   VCtl.FirstPhNum = VCtl.ResPhase;
%   VCtl.SecondPhNum = VCtl.SliceNum;
% Others
VCtl.FlipAng=90; % degree
VCtl.TEAnchorTime=0;
%Simuh.ISO = each dimension (Simuh.ISO's) *2 - 2
VCtl.ISO=[46 55 46];
VCtl.CS=0;%VObj.ChemShift*VCtl.B0;
VObj.SpinNum=1; % Controllable Spin number in each voxel
                                % Transfer VCtl to VObj, should have better
                                % way to do this ??


VCtl.TRNum=VCtl.FirstPhNum*VCtl.SecondPhNum;
VCtl.FreqDir='A/P';

%% VMag Virtual Magnetic Field
Mxdims=size(VObj.Rho);
if numel(Mxdims)==2
    if Mxdims(1)==1 | Mxdims(2)==1
        Mxdims(1)=1;
        Mxdims(2)=1;
        Mxdims(3)=1;
    end
end
VMag=struct(                                      ...
           'FRange',    ones([Mxdims(1), Mxdims(2), Mxdims(3)]), ...
           'dB0',       zeros([Mxdims(1), Mxdims(2), Mxdims(3)]),    ...
           'dWRnd',     zeros([Mxdims(1), Mxdims(2), Mxdims(3), VObj.SpinNum, VObj.TypeNum]), ...
           'Gzgrid',    zeros([Mxdims(1), Mxdims(2), Mxdims(3)]), ...
           'Gygrid',    zeros([Mxdims(1), Mxdims(2), Mxdims(3)]), ...
           'Gxgrid',    zeros([Mxdims(1), Mxdims(2), Mxdims(3)]) ...
          );

% Gradient Grid             

    [VMgd.xgrid,VMgd.ygrid,VMgd.zgrid]=meshgrid((-VCtl.ISO(1)+1)*VObj.XDimRes:VObj.XDimRes:(VObj.XDim-VCtl.ISO(1))*VObj.XDimRes,...
                                                (-VCtl.ISO(2)+1)*VObj.YDimRes:VObj.YDimRes:(VObj.YDim-VCtl.ISO(2))*VObj.YDimRes,...
                                                (-VCtl.ISO(3)+1)*VObj.ZDimRes:VObj.ZDimRes:(VObj.ZDim-VCtl.ISO(3))*VObj.ZDimRes); 
    
    [pathstr,name,ext]=fileparts('C:\MRiLab-GUILess\Grad_LinearHead.XML');
    eval(['[GxR,GyPE,GzSS]=' name ';']);
    
    % Calculate G*grid based on gradient profile, gradient integral
    if isempty(find(GxR ~=0, 1))
        Gxgrid = VMgd.xgrid;
    else
        TmpGxR=GxR(:,:,:,1);
        TmpGxR(VMgd.xgrid<=0) = 0;
        Gxgrid = cumsum(TmpGxR,2) .* VObj.XDimRes;
        TmpGxR=GxR(:,:,:,1);
        TmpGxR(VMgd.xgrid>=0) = 0;
        Gxgrid = Gxgrid + flipdim(cumsum(flipdim(-TmpGxR,2),2),2) .* VObj.XDimRes;
        
        TmpGxR=GxR(:,:,:,2);
        TmpGxR(VMgd.ygrid<=0) = 0;
        Gxgrid = Gxgrid + cumsum(TmpGxR,1).* VObj.YDimRes;
        TmpGxR=GxR(:,:,:,2);
        TmpGxR(VMgd.ygrid>=0) = 0;
        Gxgrid = Gxgrid + flipdim(cumsum(flipdim(-TmpGxR,1),1),1) .* VObj.YDimRes;
        
        TmpGxR=GxR(:,:,:,3);
        TmpGxR(VMgd.zgrid<=0) = 0;
        Gxgrid = Gxgrid + cumsum(TmpGxR,3) .* VObj.ZDimRes;
        TmpGxR=GxR(:,:,:,3);
        TmpGxR(VMgd.zgrid>=0) = 0;
        Gxgrid = Gxgrid + flipdim(cumsum(flipdim(-TmpGxR,3),3),3) .* VObj.ZDimRes;
        
    end
    
    if isempty(find(GyPE ~=0, 1))
        Gygrid = VMgd.ygrid;
        
    else
        TmpGyPE=GyPE(:,:,:,1);
        TmpGyPE(VMgd.xgrid<=0) = 0;
        Gygrid = cumsum(TmpGyPE,2) .* VObj.XDimRes;
        TmpGyPE=GyPE(:,:,:,1);
        TmpGyPE(VMgd.xgrid>=0) = 0;
        Gygrid = Gygrid + flipdim(cumsum(flipdim(-TmpGyPE,2),2),2) .* VObj.XDimRes;
        
        TmpGyPE=GyPE(:,:,:,2);
        TmpGyPE(VMgd.ygrid<=0) = 0;
        Gygrid = Gygrid + cumsum(TmpGyPE,1) .* VObj.YDimRes;
        TmpGyPE=GyPE(:,:,:,2);
        TmpGyPE(VMgd.ygrid>=0) = 0;
        Gygrid = Gygrid + flipdim(cumsum(flipdim(-TmpGyPE,1),1),1) .* VObj.YDimRes;
        
        TmpGyPE=GyPE(:,:,:,3);
        TmpGyPE(VMgd.zgrid<=0) = 0;
        Gygrid = Gygrid + cumsum(TmpGyPE,3) .* VObj.ZDimRes;
        TmpGyPE=GyPE(:,:,:,3);
        TmpGyPE(VMgd.zgrid>=0) = 0;
        Gygrid = Gygrid + flipdim(cumsum(flipdim(-TmpGyPE,3),3),3) .* VObj.ZDimRes;
        
    end
    
    if isempty(find(GzSS ~=0, 1))
        Gzgrid = VMgd.zgrid;
        
    else
        TmpGzSS=GzSS(:,:,:,1);
        TmpGzSS(VMgd.xgrid<=0) = 0;
        Gzgrid = cumsum(TmpGzSS,2) .* VObj.XDimRes;
        TmpGzSS=GzSS(:,:,:,1);
        TmpGzSS(VMgd.xgrid>=0) = 0;
        Gzgrid = Gzgrid + flipdim(cumsum(flipdim(-TmpGzSS,2),2),2) .* VObj.XDimRes;
        
        TmpGzSS=GzSS(:,:,:,2);
        TmpGzSS(VMgd.ygrid<=0) = 0;
        Gzgrid = Gzgrid + cumsum(TmpGzSS,1) .* VObj.YDimRes;
        TmpGzSS=GzSS(:,:,:,2);
        TmpGzSS(VMgd.ygrid>=0) = 0;
        Gzgrid = Gzgrid + flipdim(cumsum(flipdim(-TmpGzSS,1),1),1) .* VObj.YDimRes;
        
        TmpGzSS=GzSS(:,:,:,3);
        TmpGzSS(VMgd.zgrid<=0) = 0;
        Gzgrid = Gzgrid + cumsum(TmpGzSS,3) .* VObj.ZDimRes;
        TmpGzSS=GzSS(:,:,:,3);
        TmpGzSS(VMgd.zgrid>=0) = 0;
        Gzgrid = Gzgrid + flipdim(cumsum(flipdim(-TmpGzSS,3),3),3) .* VObj.ZDimRes;
        
    end
AP1=5;
AP2=105;
%The scan process parameter for FreqDir is A/P, which corresponds to the
%Axial view. (Anterior-Posterior)
%The NoFreqAlias, NoPhaseAlias, NoSliceAlias fields of VCtl have not yet
%been created.
     VMag.Gxgrid=Gygrid;
     VMag.Gygrid=Gxgrid;
     VMag.Gzgrid=Gzgrid;
            
     VMag.FRange(1:AP1-1,:,:)=0;
     VMag.FRange(AP2:end,:,:)=0;
if VObj.SpinNum >1
    InddWRnd=linspace(0.01,0.99,VObj.SpinNum);
    for j=1:VObj.TypeNum
        for i=1:VObj.SpinNum
            VMag.dWRnd(:,:,:,i,j)=(1./VObj.T2Star(:,:,:,j)-1./VObj.T2(:,:,:,j)).*tan(pi.*(InddWRnd(i)-1/2));
            % need large number of spins for stimulating T2* effect,
            % insufficient number of spins may cause in-accurate simulation
        end
    end
end

    [VMmg.xgrid,VMmg.ygrid,VMmg.zgrid]=meshgrid((-(Mxdims(2)-1)/2)*VObj.XDimRes:VObj.XDimRes:((Mxdims(2)-1)/2)*VObj.XDimRes,...
                                                (-(Mxdims(1)-1)/2)*VObj.YDimRes:VObj.YDimRes:((Mxdims(1)-1)/2)*VObj.YDimRes,...
                                                (-(Mxdims(3)-1)/2)*VObj.ZDimRes:VObj.ZDimRes:((Mxdims(3)-1)/2)*VObj.ZDimRes);
    [pathstr,name,ext]=fileparts('C:\MRiLab-GUILess\Mag_LinearHead.XML');
    eval(['dB0=' name ';']);
    VMag.dB0=dB0;

%% VObj Virtual Object
VObj.Mx=zeros([Mxdims(1), Mxdims(2), Mxdims(3), VObj.SpinNum, VObj.TypeNum]);
VObj.My=zeros([Mxdims(1), Mxdims(2), Mxdims(3), VObj.SpinNum, VObj.TypeNum]);
VObj.Mz=zeros([Mxdims(1), Mxdims(2), Mxdims(3), VObj.SpinNum, VObj.TypeNum]);
for i=1:VObj.TypeNum
    VObj.Mz(:,:,:,:,i)=repmat(double(VObj.Rho(:,:,:,i))/double(VObj.SpinNum),[1,1,1,VObj.SpinNum]);
end

%% VSig Virtual Signal
VSig.Mx=VObj.Mx;
VSig.My=VObj.My;
VSig.Mz=VObj.Mz;
VSig.Muts=0;

%% VCoi Virtual Coils
% B1Level: linear scale factor for B1. The input B1+ field with magnitude of this number produces nominal flip angle
% E1Level: linear scale factor for E1. The input E1+ field is scaled by an factor of nominal rf amplitude normalzed by this number
VCoi=struct( ...
            'TxCoilNum', 1, ...
            'RxCoilNum', 1, ...
            'TxCoilDefault',1,... % use default Tx Coil
            'RxCoilDefault',1,... % use default Rx Coil
            'TxCoilmg',ones([Mxdims(1), Mxdims(2), Mxdims(3)]), ...
            'TxCoilpe',zeros([Mxdims(1), Mxdims(2), Mxdims(3)]), ...
            'RxCoilx',ones([Mxdims(1), Mxdims(2), Mxdims(3)]), ...
            'RxCoily',zeros([Mxdims(1), Mxdims(2), Mxdims(3)]), ...
            'TxE1x',0,...
            'TxE1y',0,...
            'TxE1z',0,...
            'RxE1x',0,...
            'RxE1y',0,...
            'RxE1z',0 ...
            );
%Transmitter coil code
    [VMco.xgrid,VMco.ygrid,VMco.zgrid]=meshgrid((-(Mxdims(2)-1)/2)*VObj.XDimRes:VObj.XDimRes:((Mxdims(2)-1)/2)*VObj.XDimRes,...
                                                (-(Mxdims(1)-1)/2)*VObj.YDimRes:VObj.YDimRes:((Mxdims(1)-1)/2)*VObj.YDimRes,...
                                                (-(Mxdims(3)-1)/2)*VObj.ZDimRes:VObj.ZDimRes:((Mxdims(3)-1)/2)*VObj.ZDimRes);
    
    VMco.xgrid(repmat(VMag.FRange,[1,1,1,VCoi.RxCoilNum])==0)=[];
    VMco.ygrid(repmat(VMag.FRange,[1,1,1,VCoi.RxCoilNum])==0)=[];
    VMco.zgrid(repmat(VMag.FRange,[1,1,1,VCoi.RxCoilNum])==0)=[];
    VMco.xgrid=reshape(VMco.xgrid,[max(max(sum(VMag.FRange,1))),max(max(sum(VMag.FRange,2))),max(max(sum(VMag.FRange,3))),VCoi.RxCoilNum]);
    VMco.ygrid=reshape(VMco.ygrid,[max(max(sum(VMag.FRange,1))),max(max(sum(VMag.FRange,2))),max(max(sum(VMag.FRange,3))),VCoi.RxCoilNum]);
    VMco.zgrid=reshape(VMco.zgrid,[max(max(sum(VMag.FRange,1))),max(max(sum(VMag.FRange,2))),max(max(sum(VMag.FRange,3))),VCoi.RxCoilNum]);
    
    [pathstr,name,ext]=fileparts('C:\MRiLab-GUILess\Coil_8ChHead.xml');
    eval(['[B1x, B1y, B1z, E1x, E1y, E1z, Pos]=' name ';']);
    VCtl.B1Level = 0.000001; %Hard-coded this value for sequence = PSD_GRE3D, no motion, gradient = linear head, magnet = linear
    %head and coil = 8 ch for both receiver and transmitter    
    VCoi.TxCoilmg=abs(sum(B1x+1i*B1y,4))./VCtl.B1Level; % total B+ field magnitude after normalization
    VCoi.TxCoilpe=angle(sum(B1x+1i*B1y,4)); % total B+ field phase
    VCoi.TxCoilNum=1;
    VCtl.E1Level = 0.000001;
    VCoi.TxE1x=sum(E1x,4)./VCtl.E1Level; % E+ field x component after normalization
    VCoi.TxE1y=sum(E1y,4)./VCtl.E1Level; % E+ field y component after normalization
    VCoi.TxE1z=sum(E1z,4)./VCtl.E1Level; % E+ field z component after normalization
    
    VCoi.TxCoilDefault = 0;


%Create an array (empty). The first time somebody selects Tx and a certain
%coil type, make two new entries (xml file and xml file directory) and
%store the values. Same thing for Rx. Afterwards, if somebody switches
%between Tx and Rx, don't add new fieldnames. Instead simply access the
%existing fieldnames and change the values within. 

    [pathstr,name,ext]=fileparts('C:\MRiLab-GUILess\Coil_8ChHead.xml');
    eval(['[B1x, B1y, B1z, E1x, E1y, E1z, Pos]=' name ';']);
    VCoi.RxCoilx=B1x./(1e-6); % B- field x component after normalization
    VCoi.RxCoily=B1y./(1e-6); % B- field y component after normalization
    VCoi.RxCoilNum=length(Pos(:,1));
    VCoi.RxCoilDefault = 0;
    VCoi.RxE1x=E1x./(1e-6); % E- field x component after normalization
    VCoi.RxE1y=E1y./(1e-6); % E- field y component after normalization
    VCoi.RxE1z=E1z./(1e-6); % E- field z component after normalization

%% VVar Virtual Pulse Packet Initialization
VVar=struct(                          ...
    'rfAmp',        zeros(VCoi.TxCoilNum, 1), ...
    'rfPhase',      zeros(VCoi.TxCoilNum, 1), ...
    'rfFreq',       zeros(VCoi.TxCoilNum, 1), ...
    'rfCoil',       0, ...
    'rfRef',        0,              ...
    'GzAmp',        0,              ...
    'GyAmp',        0,              ...
    'GxAmp',        0,              ...
    'ADC',          0,              ...
    'Ext',          0,              ...
    't',            0,              ...
    'dt',           0,              ...
    'rfi',          0,              ...
    'Gzi',          0,              ...
    'Gyi',          0,              ...
    'Gxi',          0,              ...
    'ADCi',         0,              ...
    'Exti',         0,              ...
    'utsi',         0,              ...
    'Kz',           0,              ...
    'Ky',           0,              ...
    'Kx',           0,              ...
    'SliceCount',   0,              ...
    'PhaseCount',   0,              ...
    'TRCount',      0,              ...
    'ObjLoc',       [0;0;0],        ...  % Object location
    'ObjTurnLoc',   [0;0;0],        ...  % Object turning point location
    'ObjAng',       0,              ...  % Object rotating angle
    'ObjMotInd',    0,              ...  % Object motion section index
    'gpuFetch',     0               ...  % Flag for fetching GPU data at extended process 
    );

end

File: PreScan.py
#Need numpy : pip install numpy
#Need LoadImage, VObj in it

# from LoadImage import VObj
import array
import numpy as np

#   VCtl
#These are the calculations
#   VCtl.ResFreq = VCtl.ResFreq + ~mod(VCtl.ResFreq,2) % guarantee odd number of Kx sample points for sampling echo peak when Kx = 0
#   VCtl.RFreq=VCtl.FOVFreq/(VCtl.ResFreq - 1);
#   VCtl.ResPhase = VCtl.ResPhase - mod(VCtl.ResPhase,2); % guarantee even number of Ky sample points for Ky = 0
#   VCtl.RPhase=VCtl.FOVPhase/VCtl.ResPhase;
#   VCtl.SliceNum = max(1,VCtl.SliceNum - mod(VCtl.SliceNum,2)); % guarantee even number of Kz sample points for Kz = 0
#   VCtl.RSlice=VCtl.SliceThick;
#   VCtl.FOVSlice=VCtl.SliceNum*VCtl.SliceThick;
#   VCtl.TrajType='cartesian';
#   VCtl.FirstPhNum = VCtl.ResPhase;
#   VCtl.SecondPhNum = VCtl.SliceNum;
#   Simuh.ISO = each dimension (Simuh.ISO's) *2 - 2

#How to make VCtlglobal? if not : array?
VCtl ={
    "ResFreq": 101,
    "RFreq": 0.002,
    "ResPhase": 80,
    "RPhase": 0.0025,
    "SliceNum": 1,
    "RSlice": 0.006,
    "FOVSlice": 0.006,
    "TrajType": "cartesian",
    "FirstPhNum": 80,
    "SecondPhNum": 1,
    "FlipAng": 90, # degree
    "TEAnchorTime": 0,
    "SpinNum": 1,
    "TRNum": 80,
    "FreqDir": "A/P",
    #to be changed for custom phantoms :
    "ISO": [46, 55, 46], #ISO is a List
    "CS": 0,# =ChemShift*VCtl.B0
}
#MxDims=size(VObj.Rho)
Mxdims=np.array([90, 108, 90]) #by default is a row matrix

#Needed for VMag :
VObj ={
#all hardcoded
    "SpinNum" : 1,
    "TypeNum" : 1,
    "XDimRes" : 0.002,
    "YDimRes" : 0.002,
    "ZDimRes" : 0.002,
    "XDim" : 90,
    "YDim" : 108,
    "ZDim" : 90,
}
#VMag Virtual Magnetic Field
VMag ={
#Syntax: np.ones(number of pages, rows, cols)
    "FRange": np.ones((Mxdims[0], Mxdims[1], Mxdims[2])),
    "dB0" : np.zeros((Mxdims[0], Mxdims[1], Mxdims[2])),
    "dWRnd" : np.zeros((Mxdims[0], Mxdims[1], Mxdims[2], VObj["SpinNum"], VObj["TypeNum"])),
    "Gzgrid" : np.zeros((Mxdims[0], Mxdims[1], Mxdims[2])),
    "Gygrid" : np.zeros((Mxdims[0], Mxdims[1], Mxdims[2])),
    "Gxgrid" : np.zeros((Mxdims[0], Mxdims[1], Mxdims[2])),
}

#Gradient Grid
ISO=VCtl["ISO"]

a=(-ISO[0]+1)*VObj["XDimRes"]
b=(VObj["XDim"]-ISO[0])*VObj["XDimRes"]
c=VObj["XDimRes"]
#arange is used because some parameters are floats and not ints. Range only takes in integers
Gxgrid, Gygrid, Gzgrid = np.meshgrid(np.arange(a, b, c ),
np.arange((-ISO[1]+1)*VObj["YDimRes"], (VObj["YDim"]-ISO[1])*VObj["YDimRes"], VObj["YDimRes"]),\
np.arange((-ISO[2]+1)*VObj["ZDimRes"], (VObj["ZDim"]-ISO[2])*VObj["ZDimRes"], VObj["ZDimRes"]))

AP1=5
AP2=105

LR1 = 6
LR2 = 86

SI1 = 45
SI2 = 48
# The scan process parameter for FreqDir is A/P, which corresponds to the
# axial view. (Anterior-Posterior)
# The NoFreqAlias, NoPhaseAlias, NoSliceAlias fields of VCtl have not yet
# been created.
VMag["Gxgrid"]=Gygrid
VMag["Gygrid"]=Gxgrid
VMag["Gzgrid"]=Gzgrid

FRange = VMag["FRange"]
FRange[range(0, AP1-1), :, :]=0
#indexing differences in Python vs Matlab
#python :  range(start, end) does not include end
#So add 1 to ZDim, YDim??? I get an error for Y and Z, not for X.
FRange[range(AP2,VObj["XDim"]), :, :]=0

FRange[:, range(0, LR1-1) ,:]=0
FRange[:, range(LR2,VObj["YDim"]) ,:]=0

FRange[:, :, range(0,SI1-1)]=0
FRange[:, :, range(SI2,VObj["ZDim"])]=0

VObj["Mx"] = np.zeros((Mxdims[0], Mxdims[1], Mxdims[2], VObj["SpinNum"], VObj["TypeNum"]))
VObj["My"] = np.zeros((Mxdims[0], Mxdims[1], Mxdims[2], VObj["SpinNum"], VObj["TypeNum"]))
VObj["Mz"] = np.zeros((Mxdims[0], Mxdims[1], Mxdims[2], VObj["SpinNum"], VObj["TypeNum"]))

VObj_Mz=VObj["Mz"]
VObj_Rho=VObj["Rho"]
for i in range(VObj["TypeNum"]):
    VObj_Mz[:,:,:,:,i]=np.tile(double(VObj_Rho[:,:,:,i])/double(VObj["SpinNum"]) , [1,1,1,VObj["SpinNum"]])

#VSig Virtual Signal
VSig ={
    "Mx": VObj["Mx"],
    "My": VObj["My"],
    "Mz": VObj["Mz"],
    "Muts" : 0,
}

 #    VCoi Virtual Coils
 # B1Level: linear scale factor for B1. The input B1+ field with magnitude of this number produces nominal flip angle
 # E1Level: linear scale factor for E1. The input E1+ field is scaled by an factor of nominal rf amplitude normalzed by this number
VCoi ={
    'TxCoilNum': 1,
    'RxCoilNum': 1,
    'TxCoilDefault': 1, #use default Tx Coil
    'RxCoilDefault': 1, #% use default Rx Coil
    'TxCoilmg': np.ones((Mxdims[0], Mxdims[1], Mxdims[2])),
    'TxCoilpe': np.zeros((Mxdims[0], Mxdims[1], Mxdims[2])),
    'RxCoilx': np.ones((Mxdims[0], Mxdims[1], Mxdims[2])),
    'RxCoily': np.zeros((Mxdims[0], Mxdims[1], Mxdims[2])),
    'TxE1x': 0,
    'TxE1y': 0,
    'TxE1z': 0,
    'RxE1x': 0,
    'RxE1y': 0,
    'RxE1z': 0,
}

# VMot Virtual Motion
VMot={
    't': 0,
    'ind': 1,
    'Disp': np.array([[0],[0],[0]]),
    'Axis': np.array([[1],[0],[0]]),
    'Ang': 0,
}

# VVar Virtual Pulse Packet Initialization
VVar={
    'rfAmp': np.zeros((VCoi["TxCoilNum"], 1)),
    'rfPhase': np.zeros((VCoi["TxCoilNum"], 1)),
    'rfFreq': np.zeros((VCoi["TxCoilNum"], 1)),
    'rfCoil': 0,
    'rfRef': 0,
    'GzAmp': 0,
    'GyAmp': 0,
    'GxAmp': 0,
    'ADC': 0,
    'Ext': 0,
    't': 0,
    'dt': 0,
    'rfi': 0,
    'Gzi': 0,
    'Gyi': 0,
    'Gxi': 0,
    'ADCi': 0,
    'Exti': 0,
    'utsi': 0,
    'Kz': 0,
    'Ky': 0,
    'Kx': 0,
    'SliceCount': 0,
    'PhaseCount': 0,
    'TRCount': 0,
    'ObjLoc': np.array([[0],[0],[0]]),  # Object location
    'ObjTurnLoc': np.array([[0],[0],[0]]),  # Object turning point location
    'ObjAng': 0,  # Object rotating angle
    'ObjMotInd': 0,  # Object motion section index
    'gpuFetch': 0,  # Flag for fetching GPU data at extended process
}
